### DISCUSSION TITLE
A quick note on breaking changes

### DISCUSSION THREAD

**Discussion by lthoerner** [ratatui/ratatui#163] [Discussion]
Hey guys, I just wanted to share some quick thoughts about breaking changes to the crate.

I think that we may be approaching making breaking changes with too much caution; it seems to be the case that the maintainers are loathe to implement them, which is entirely understandable. Breaking changes can make it a pain for users of the crate to keep up with releases if they want access to new features.

However, I believe that one of the important hallmarks of a pre-major-version (you could call it "beta" if applicable) project, is the ability to "move fast and break things." Of course, it should always be considered whether a feature is really necessary, but we've been left with a bit of an underbaked cake from the original `tui-rs` developers (no offense at all, we've all had projects we didn't truly finish) - and I think that it is often necessary for us to patch existing functionality or add new functionality in ways that are not conducive to backward compatibility.

Overall, my point is that programmers who are not only using a beta crate, but also expecting to frequently update it for new features, should be ready to do revisions to their own code. And, if they do not wish to do so, they can simply not update the crate. Until we release `ratatui 1.0.0`, I don't think there is any point in being overly devoted to backward compatibility.

I'm looking forward to hearing you guys' thoughts on this. I might be entirely off-base here, but I wanted to bring it up.

**Comment by joshka** [comment]
+1 I just wrote something related in https://github.com/tui-rs-revival/ratatui/pull/122 regarding whether to add new methods to fix the places where we consume widget rather than borrowing it (and which probably applies throughout the entire library for things where we should use borrow or use `Cow`.

>One possible other approach to implementing a new method would to actually break callers by making render_widget and render_stateful_widget actually borrow the widget (and probably area too) instead of having two methods. We're in a spot where it could be reasonable to make a choice to break consumers here if the outcome enables many more scenarios and makes the library more ergonomic - this is probably controversial and I'd defer to those who have more experience than I do on this.

**Comment by sophacles** [comment]
I'm not particularly opposed to breaking changes, but we should at least note what they are and just how breaking they are.

Something like the addition of an enum variant is far less disruptive than say a major refactor that changes how to construct an app. Even if the users are expecting breaking changes it's nice to at least tell them what we broke so they don't have to stare at diffs to track it down.

I also think major changes should involve some discussion and feedback time from users - a couple reasons for this:

1. If major changes are going to happen, it's nice to find out if the users actually want that, or if they have other related changes that can be bundled into the same release to reduce downstream toil.

2. Major changes should be planned - it's easy to make big changes that seem convenient or useful but paint the library into a corner or reduce its broad applicability. 

This library existed in stasis for a long time. Independent of some semantic versioning argument,  lots of software exists for decades in a relatively stable state with a <1 version number. When it suddenly starts dropping breaking changes on a regular basis, it's annoying (at least, sometimes downright infuriating). I think it's worth being deliberate (not pausing breaking changes, just being deliberate) for a while to respect our users that might be surprised by the new revival. I know they have to make a conscious choice to switch from tui-rs to ratatui, but that change does not imply a sudden change in development cycles/patterns.

**Reply by lthoerner** [reply]
Of course. They should be well-documented.

I do contend a bit with the idea that having a <1 version number is not particularly meaningful. I don't believe you can extricate the meaning of a version number from the number itself. I understand that some software ignores SV but if we are purporting to follow it, we must be careful not to diverge too much from the expected significance of versioning. If we are going to make the claim that `ratatui` is stable and no longer in beta (which it almost certainly is not), we need to release 1.0.0. Until then I think that even major changes are to be expected. Again, users who find the software to be stable as-is, and/or who feel as though it meets their needs already, can simply not update the crate.

You are right in saying that this can be annoying for users. But sometimes you just have to rip the band-aid off. Again, this doesn't have to happen often. But I don't think we need to go through weeks of community feedback for every minorly-breaking PR. We need a plan for the way we are going to approach this rather than just letting the chips fall where they may.

**Reply by joshka** [reply]
Breaking changes act as a barrier to upgrade (35 dependents), or migration from tui to ratatui (350 dependents). The major concern is how much work has to be done elsewhere to support the work done here.
That work is not huge, but it's meaningful and we owe it to the users to make it small and easy.

Because of that, we should break only when necessary, but in doing so understand the impact (how many users / potential migrations are broken and what do they have to do to fix it). It's not a bandaid / not bandaid thing, it's more of needing to assess whether the pain of fixing is more than the pain of leaving it alone.

For those reasons, I'd say we just need to come up with a set of criteria that we mark of on when to break consumers and stick with that for any change. Off the top of my head, the things I'd say should be on this checklist are:

1. Do we have a good reason for the breaking change?
2. Have we considered non-breaking changes to fix the issue?
3. Have we documented what the change is in detail before making it?
4. Are there other changes around the same area that would reduce the fatigue of responding to updates?
5. Have we documented how to fix the change in downstream code?
6. Do we know how many dependents / migrations use the code being changed?
7. Do we have a plan for communicating the change (and a timeline)?
8. Do we need user feedback on this change?
9. Should we wait a little for other things to settle before making this change?
10. What else?

Like many rust projects we're not 1.0, which means we can be more lax with semver, but just because we can, doesn't mean we have to. Having a plan doesn't have to be a huge blocker, but there has to be one.

I don't agree that we're being overly cautious on making changes. I just believe that we haven't particularly put the time into deciding how breaking changes should be handled. That doesn't have to be too difficult, but it does have to be done before breaking changes happen.

**Reply by lthoerner** [reply]
I agree with basically all of this. However, the idea that we must maintain drop-in compatibility with `tui-rs` is incredibly limiting, so I don't think this should be a major focus going forward. We should, of course, document changes from the original library, and offer a note of the last version number where the library was still a drop-in replacement. But overall it's going to be close to impossible to keep them totally interchangeable.

My only major concern about this list is that it is quite exhaustive, so we will need more people working on the library to investigate these things (or set up automation on some of it) before it is feasible to do the research on each change we need to make.

**Reply by lthoerner** [reply]
To clarify, I'm not saying that you claimed we needed to maintain interoperability - just mentioning because I've seen it talked about before, and I wanted to make note of it.

**Reply by joshka** [reply]
>the idea that we must maintain drop-in compatibility with tui-rs is incredibly limiting

Yep, those limitations are intentional though. The idea really follows from the higher order concerns:
1. There are 350 existing tui projects that haven't yet migrated to ratatui. Some won't ever, but each breaking change increases the friction of making that happen.
2. Every breaking change introduces an amount of undifferentiated downstream work. It's important to have some empathy for the users, and let that empathy find some balance with implementing features. Rust seems like a language where that sort of approach is regularly seen as a positive aspect - people value the stability of the language, libraries etc.

>My only major concern about this list is that it is quite exhaustive, so we will need more people working on the library to investigate these things (or set up automation on some of it) before it is feasible to do the research on each change we need to make.

That list is a purely a product of a few minutes brainstorm, and I'd welcome feedback if you'd like to suggest some. What would you add? What would you take away? What would you modify?

Filling it out didn't take more than a minute or two on #172, and didn't really have any effect on the delivery of the feature other than to perhaps consider some things that might have been missed. But that minute or two might save many hours of downstream work. That's a pretty huge difference that to me makes it well worth investing the time.

**Reply by lthoerner** [reply]
I'm curious as to how many breaking changes have been made since forking from `tui-rs`, if any. That would give me a better idea of the relative importance of drop-in compatibility.

**Reply by joshka** [reply]
> I'm curious as to how many breaking changes have been made since forking from `tui-rs`, if any. That would give me a better idea of the relative importance of drop-in compatibility.

Only one so far:
- https://github.com/tui-rs-revival/ratatui/pull/42 - fixes a bug causing rendering to changes for people using it
And 3 that are merged but not released yet:
- https://github.com/tui-rs-revival/ratatui/pull/114 - I don't know about this one
- https://github.com/tui-rs-revival/ratatui/pull/168 - introduces a new type, which can break implicit conversions
- https://github.com/tui-rs-revival/ratatui/pull/133 - fixes a bug, rendering changes for people using the bug

**Reply by lthoerner** [reply]
What is our plan going forward on breaking the API? It eventually will need to happen. My opinion is that we provide consumers with a record of the last version number before breaking changes occurred (where the crate can still be a drop-in replacement for `tui-rs`) and then thoroughly document the requirements for updating to the latest version as we make modifications to the API.

**Reply by sayanarijit** [reply]
IMO breaking changes are fine if it's the only way to fix a bug. Minor breaking changes are also fine, like changing function parameters to accept references instead of owned values to optimize performance. But for renames and structural changes, I'd keep a backward compatible alias, or atleast move the legacy code to a module dedicated to maintain compatibility, like when serde_yaml broke compatibility, in 0.8->0.9, the author just [moved the old logic to the `with` module](https://github.com/dtolnay/serde-yaml/commit/f344c5643c616977472560a8c6d8b07509b3045e). And [here's how I do it](https://github.com/sayanarijit/xplr/blob/dev/src/compat.rs).

The point is, let's try not to break the API as much as possible, and when we do have to break things, let's atleast give the users some easy fallback option, with proper documentation and upgrade instructions.

**Reply by lthoerner** [reply]
In some cases it will be close to impossible to maintain backwards compatibility without making the codebase a bit of a mess (see our discussion on #175 about maintaining compatibility with unnamed fields). I agree with your last point, but when it needs to happen, I think it's totally fine as long as it's well-documented for the users (I think that being very reluctant to make changes is somewhat unnecessary).

At the end of the day, if `tui-rs` had a great existing API I wouldn't be so gung-ho to revamp/refactor it in major ways but it has quite a few problems that can only truly be fixed by breaking changes. What is the purpose of `ratatui` if not to fix the pain points created by `tui-rs`?

**Reply by sayanarijit** [reply]
> see our discussion on https://github.com/tui-rs-revival/ratatui/pull/175 about maintaining compatibility with unnamed fields

I think unnamed fields shouldn't be public. But if it's already public, it should be left as it is. [We can keep both: compatibility and a clean codebase by moving things here and there a bit](https://github.com/tui-rs-revival/ratatui/pull/175#issuecomment-1546066219).

> What is the purpose of ratatui if not to fix the pain points created by tui-rs?

As the readme says, we want to continue development by building upon the foundation already laid by the original author, not a complete rewrite/refactor of the tui crate. Still I believe there's potential to go a long way even if we keep breaking changes to a minimum.

**Reply by lthoerner** [reply]
If we were talking about this after a 1.0 release I would tend to agree with you, but we're not, so I think we should be fixing bad design rather than accommodating it. You can only make so many additive changes without modifying the underlying API to an extent.

**Reply by sayanarijit** [reply]
That actually is a valid point. But majority of the rust crates are <1.0, yet the community try to maintain compatibility as much as possible. So adding frequent and huge breaking changes, using this excuse, without considering the community opinions might not be well received. Also, I guess we're not too far from 1.0 considering the ecosystem we already have. The only thing is we have to wait until all our dependencies are stable enough.

**Reply by lthoerner** [reply]
My opinion as a user of the crate is that we are a long way from 1.0 - but that is largely up to the maintainers. I understand that a lot of crates don't follow semver closely and I've actually seen this mentioned on r/rust a few times; I think that the relative novelty of Rust as a language tends to push crate developers toward "forever beta" releases. My thought on that would be that we should be doing it right rather than just following the crowd, but I can see arguments either way being compelling enough.

**Comment by Byron** [comment]
I really appreciate all the energy and excitement the new maintainers are showing here and believe this should lead to the best possible TUI library possible even if it entails breaking changes. Personally, I am happy with how my software looks like (`dua-cli`, `prodash`), but also had to workaround a few issues  `tui-react` crate. Hopefully, thanks to your work, that can go away one day.

That said, I'd be happy to wait until 1.0 even and then upgrade everything at once, which should be straightforward as there is excellent documentation that lets me search for old names and trace forward to how its supposed to be used now. Doing this in little steps might be easier, but I'd only do so if forced (for instance, by CVEs or other issues due to outdated dependencies). Right now, I'd love to just keep things working for as long as possible.

That said, there might be an opportunity here as well, as you could validate breaking changes against the top-list of your users. At the very least that would give them the upgrade to the next version for free, and at best you get to create even better/more convenient APIs.

All the best!

**Reply by joshka** [reply]
P.s. @Byron's crates are the top of the tui reverse dependencies list by about 10x compared to the next closest (tui-logger, tokio-console, gitui)

**Reply by lthoerner** [reply]
I think this is the way to go - maintaining the drop-in version for the time being with essential patches, while developing 1.0.0 without too much reservation for making API updates, and thoroughly documenting the API changes so users can upgrade when the time comes to make a major version release. I think that it just isn't feasible to make only additive changes to a product that was abandoned halfway through.

Happy to hear from the developers using `ratatui` in their own projects. Thanks for the feedback!

**Reply by mindoodoo** [reply]
> maintaining the drop-in version for the time being with essential patches, while developing 1.0.0 without too much reservation for making API updates

I like this idea, however I think if we were to take this approach we would need to be clear with what we really want to do for 1.0.0, instead of just feeling more free to introduce breaking changes. I don't believe we should develop two versions side by side forever which is why we should know precisely what we're striving for if we ever start work on 1.0.0 while maintaining the available version.

But all in all, if we ever reach a point where we have a clear idea of what changes we would like to implement for 1.0.0, I agree with @Eyesonjune18 in saying this would be a great way to do it.

**Reply by lthoerner** [reply]
Well yes, but also keep in mind that requirements will change over time. We won't necessarily know all of our end goals before starting development. But I'm definitely open to hearing roadmap ideas.

**Comment by lthoerner** [comment]
One thing I thought to mention is - even entire *programming languages* do frequent breaking releases. How many apps use Java, for instance? And of those, how many are still on Java 8 because it works for their needs? And then you have languages like C++ who sacrifice correctness and ease of use for backward compatibility, which IMO is not something we should strive for. Of course, not all languages and software projects should do release cycles like Java, but it's not like the idea of having to modify your application to use new features is a foreign one to the development world. I don't think it should require weeks-long conversations/debates every time we need to rename a struct field. To reiterate past comments, I'm confused why we're talking about this crate like it's some sort of legacy software when it hasn't even had a major version release yet.

**Reply by sayanarijit** [reply]
> I'm confused why we're talking about this crate like it's some sort of legacy software when it hasn't even had a major version release yet.

Regardless of the versioning situation, we basically have the following choices:

- Keep the tui-rs design, along with its existing ecosystem and user base.
- Make a drastically different (maybe potentially better) design, abandoning the existing ecosystem and user base.

Generally on most decisions like this, we can't have it both ways, but in this case, there's a third option with some compromises:

- Gradually improve the tui-rs design, but carefully, walk along with the ecosystem and userbase, slow and steady.

If I didn't care about the tui-rs community, I'd just make a new library, and try to sell it to the users willing to migrate. In which case, the tui-rs community, who are unwilling to migrate, would still come along to revive the "legacy" library, without me. Which is totally fine, and possibly a better approach. But the fact is, we think there's no new tui library worth migrating to yet, and in this, we're the basically "unwilling" community reviving tui-rs.

**Reply by lthoerner** [reply]
I don't think that the second option requires _abandoning_ the current userbase. I think you may be thinking of this a bit too drastically. The top user of the crate has weighed in above, and it did not seem that their statement fell anywhere near saying "I will stop using Ratatui if any major changes are made."

Regarding your third option, my point all along was that we shouldn't have to take everything incredibly slow and steady. As long as we're not being outright hostile to the thirty-some people using our crate, and are able to effectively communicate changes as they happen, I see it as a waste of time to sit and wait for feedback on every single change. The issues I have with this approach can be summed up as such:
- The original design is not all that great in a lot of places. Additive-only changes aren't feasible if we want to make any major improvements.
- The userbase is neither large nor active enough for us to get quick and meaningful feedback on most features, especially minor ones. Requiring user review for even minor changes seems like more of a burden than a boon to users.

See my comments above for a more complete idea of the approach I think we should take.

**Reply by sayanarijit** [reply]
So yes, in your example, we're the c++ community, trying to maintain it, due to either - rust hasn't been released yet, or we don't think rust it worth migrating to, or it's not possible to migrate, or any other reason you'd think of. It doesn't really matter if it's c++ v1 or v0.1.

> Edit: sorry for the redundant comment.

**Reply by sayanarijit** [reply]
> The userbase is neither large nor active enough for us to get quick and meaningful feedback on most features, especially minor ones. Requiring user review for even minor changes seems like more of a burden than a boon to users.

Here I don't agree. There's no other alternative that comes close to tui-rs ecosystem. Reg activeness, It depends of how you measure it. Different ecosystems have different level of activeness. We're definitely not as active as react or any other js framework. But compared to tui-rs alternatives, I think we're the most active community.

Update: also we don't require user review for minor changes. We just straight away merge them as long as it's not breaking.

**Reply by lthoerner** [reply]
Most active ≠ active. I wouldn't call this library particularly popular even though you are correct in saying it's probably the most popular TUI library outside of the original `tui-rs`. Just because we have a proportionally large userbase does not mean we have enough people to get meaningful feedback on every change, or even just most of them.

**Reply by sayanarijit** [reply]
We won't ask user review from all or most users, but as @joshka mentioned, we should atleast get review from users who are at the top of the reverse dependency list. And I'd add we should definitely get review from third party extensions authors, if we break their extensions. Large or small, popular or unpopular, fact is that we have users and a community, and we want to keep them.

**Comment by Byron** [comment]
I find it really hard to find the right place to chime in, and guess a new top-level reply is the best option. 

As mentioned earlier, even though having no work with `ratatui` thanks to stability is valuable, at this point at least to me it's as valuable to have regular updates and active maintainers. I do remember how slow interactions with `tui` were back in the days, and I definitely prefer the fast pace here.

Making `ratatui` the best possible TUI crate is also a valuable goal to me, as it would enable better TUIs for Rust software of all kinds. This will mean breaking changes and to me that's perfectly fine if…

* they are not done in a vacuum but on *actual users of the `tui`/`ratatui` crates*
* they won't be fully or partially undone by later updates (or at least that's the rare event, mistakes can happen)

The first point, not making breaking changes *in a vacuum*, deserves some more explanation as it to my mind gets to leverage the relatively small user-base of the `tui` crates: You can know your users and they probably cover 95% of the user-cases already. If you had a selection of crates against which you test all changes, I think the chances for these to be good are much higher with the side-effect that upgrading your key users is only a PR away. With such a workflow, users would even be able to provide early feedback on draft PRs if that's what they desire.

Of course, this also means extra effort in understanding the code-bases of these `tui` users, while trying to extract their problems and workarounds to get a first set of possible improvements to make.  As a practical example, to be able to do `dua-cli`, I felt the need for creating `tui-react` which fixes some issues I had with the `tui` API. What can be done to not need that utility crate anymore? Are there better solutions to the problem `tui-react` is solving? What's the effect of such changes to other key-users? What's the minimal viable change to remove that wart?

To me, the outlook of more active and practice-oriented development is very exciting, as I believe it can lead to Rust finally getting the TUI crate it deserves (*the best*, of course 😁).

Probably there are parts of the community that happily stay with `tui` as it works for them, and who didn't make it onto the list of privileged crates against which all changes are validated. To me, that's fine, as nothing would prevent you from submitting a PR if you think they should make the switch. Maybe some of them will make the switch by themselves because it's obvious that `ratatui` is for their benefit, and because there is a well-maintained *migration document` to help them along.

I think a lot of what I am saying here is based on my experience with `gitoxide`, which wants to eventually replace `git2` (among other things), and making upgrades for early-adopters is what I do regularly, along with the occasional 'oxidization' where I create a PR to switch to `gitoxide`. Doing this is so valuable as I get to improve the API along the way to be as good or often times better. I truly think it works.

**Comment by joshka** [comment]
The Spans -> Lines rename caused the following issues downstream:
- https://github.com/kamiyaa/joshuto/issues/321
- https://github.com/kamiyaa/joshuto/issues/318

These issues were brought to the attention of the owner by users attempting to install the tool using cargo install. It highlights that the problem we cause by breaking compatibility is not just one where we break our direct downstream developers using ratatui in their crates, but that we also break the users (or potential users) of the tools built with ratatui.

I know that apps should include a Cargo.lock and should be installed using the `--locked` flag to avoid issues like this, but we can avoid these issues too by not breaking compatibility (at least as a general rule).

**Reply by a-kenji** [reply]
The dependency was specified in the `Carg.toml` to be an exact match. Does cargo really update breaking changes according to semver on its own? Or did I miss how this happened?

**Reply by joshka** [reply]
That is odd. I'm not sure what happened there to make that pick up the 0.21 crate without the related change.
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html suggests that that shouldn't have happened.

**Reply by orhun** [reply]
I also have a similar issue in my project: https://github.com/orhun/systeroid/pull/89 (see failed pipelines)

**Reply by sayanarijit** [reply]
Cargo does update minor releases if --locked is not used. For us, 0.20 - 0.21 may be a major release, but not for cargo I guess.

**Reply by joshka** [reply]
> I also have a similar issue in my project: [orhun/systeroid#89](https://github.com/orhun/systeroid/pull/89) (see failed pipelines)

That got bumped to 0.21 as 0.21 was needed by tui-logger.

I can't work out joshuto though. Here's the tree prior to the change
```
  ~/local/migrate-to-ratatui/joshuto on  @20d51cf1                                                                 took  35s at  06:24:47
❯ cargo tree -i ratatui --all-features
ratatui v0.20.0
├── ansi-to-tui v3.0.0
│   └── joshuto v0.9.4 (/Users/joshka/local/migrate-to-ratatui/joshuto)
└── joshuto v0.9.4 (/Users/joshka/local/migrate-to-ratatui/joshuto)
```

I wonder if there's some different dependency versioning for install compared to build?

**Reply by jw013** [reply]
This appears to be a case of multiple conflicting versions. joshuto itself depended on ratutui 0.20 but its [ansi-to-tui](https://crates.io/crates/ansi-to-tui/3.0.0/dependencies) dependency declares a `"0.*"` dependency on ratatui which is how 0.21 wound up in the build. I think `"0.*"` dependencies are probably a bad idea, perhaps an issue should be filed with ansi-to-tui.

**Reply by joshka** [reply]
https://github.com/uttarayan21/ansi-to-tui/pull/31

**Comment by haydenflinner** [comment]
I'm just dropping by to comment that the only reason I'm here, using this crate, is because `cursive` is hesitant to break things by introducing lifetimes, which makes my life as a user harder than necessary. My thought is that ratatui will likely move more quickly in whatever direction the community thinks is ideal.
